### DISCUSSION TITLE
Unit Testing for components with IMessageService

### DISCUSSION THREAD

**Discussion by mdmontesinos** [Megabit/Blazorise#4303] [Q&A]
Is there any way to test a component that requires IMessageService by using bUnit?

The main issue I have is that when I render my component, the markup for it only includes the component itself, not the html code for the modal. Therefore, I'm not able to find the confirm/cancel button to keep the flow of the component going. 

I know the MessageProvider component should be included next to the App component, but how does that translate to unit testing?
Below I include the relevant code that I have so far.

Test Method:
```c#
ctx = new TestContext();
ctx.JSInterop.Mode = JSRuntimeMode.Loose;

ctx.Services
      .AddBlazorise(options =>
      {
          options.Immediate = true;
      })
      .Replace(ServiceDescriptor.Transient<IComponentActivator, ComponentActivator>())
      .AddBootstrap5Providers()
      .AddFontAwesomeIcons();

//Render Component
var cut = ctx.RenderComponent<MyComponent>();

//Act
var button = cut.Find("some button");
button.Click();

//After clicking that button, the code specified below should be executed, and the modal should open.
//Now, I need to be able to interact with the modal, to either confirm or cancel

//THIS IS WHAT I REQUIRE
var modalButton = cut.Find(".modal-footer .btn-primary");
modalButton.Click();
```

Method to be executed after clicking the button:
```c#
var confirm = await MessageService.Confirm("something");

if (confirm)
{
    //do something
}
else{
  //something else
}
```

**Comment by mdmontesinos** [comment]
I've also tried rendering the MessageProvider component in the TestContext like this:

```c#
IRenderedComponent<MessageProvider> messageProvider = ctx.RenderComponent<MessageProvider>();
```

And when I click the button that should open the modal, the content of the modal component does change (according to the title and message established). 

I'm also able to locate the Confirm/Cancel button by using the Find method in MessageProvider component. 

However, when clicking it, nothing triggers in the original component, so I'm still stuck waiting for the confirm.

Any clues on how to proceed?

**Reply by stsrki** [reply]
@David-Moreira can you help?

**Reply by David-Moreira** [reply]
Hello,
Generally what happens is that you haven't given time to re-render, you need to await the event call, or do `comp.WaitForAssertion`

This works:
![image](https://user-images.githubusercontent.com/22283161/203174756-18cdc7cb-584b-4686-aa75-84f5f19a98a2.png)

![image](https://user-images.githubusercontent.com/22283161/203174693-e9686bc1-7c95-4dd4-963e-d4417695f200.png)
![image](https://user-images.githubusercontent.com/22283161/203174704-93511875-fe22-4bc0-882a-bc9b673f7826.png)

**Reply by mdmontesinos** [reply]
The ClickAsync method did the trick and the IMessageProvider now works as expected.

Thanks for the answer!
